**Deep Learning Tutorial**, Aban 20 at 18:00 via Skype 
Class Room, DL Recitation, Session Eight

[Deep Learning](https://github.com/hhaji/Deep-Learning), Data Science Center, Shahid Beheshti University

Presented and prepared by [‌‌Yavar Yeganeh](https://github.com/YavarYeganeh)




**Acknowledgment and References:**

* [Pytorch](www.Pytorch.org)

* Others listed throughout the notebook


<table align="left">
  <td>
    <a target="_blank" href=https://colab.research.google.com/drive/1cass1LqzpmTs1dnasSzgHnvuSnHdODZQ?usp=sharing"><img src="https://colab.research.google.com/img/colab_favicon_256px.png" height="100" width="100" /> <br>  Run in Google Colab</a>
  </td>
</table>


We will discuss:

- [torchvision.models](https://pytorch.org/docs/stable/torchvision/models.html)
- Regularization for Deep Learning (Part two): Bagging and Dropout
- Optimization for Training Deep Models: Different Optimizers and Batch Normalization
- **Training** a fully connected neural networks without using machine learning libraries in Python (I'll send a tutorial for it)

A great repository for Pytorch related stuff:

* [**the-incredible-pytorch**](https://www.ritchieng.com/the-incredible-pytorch/#Optimization): https://github.com/ritchieng/the-incredible-pytorch

[**torchvision.models**]([torchvision.models](https://pytorch.org/docs/stable/torchvision/models.html)): 

The models subpackage contains definitions of models for addressing different tasks, including: image classification, pixelwise semantic segmentation, object detection, instance segmentation, person keypoint detection and video classification.

Classification: The models subpackage contains definitions for the following model architectures for image classification:

    AlexNet

    VGG

    ResNet

    SqueezeNet

    DenseNet

    Inception v3

    GoogLeNet

    ShuffleNet v2

    MobileNet v2

    ResNeXt

    Wide ResNet

    MNASNet


In [ ]:
import torch

In [ ]:
import torchvision.models as models
resnet18 = models.resnet18()
resnet18_pt = models.resnet18(pretrained=True)
#alexnet = models.alexnet()
#vgg16 = models.vgg16()
#squeezenet = models.squeezenet1_0()
#densenet = models.densenet161()
#inception = models.inception_v3()
#googlenet = models.googlenet()
#shufflenet = models.shufflenet_v2_x1_0()
#mobilenet = models.mobilenet_v2()
#resnext50_32x4d = models.resnext50_32x4d()
#wide_resnet50_2 = models.wide_resnet50_2()
#mnasnet = models.mnasnet1_0()

Downloading: "https://download.pytorch.org/models/resnet18-5c106cde.pth" to /root/.cache/torch/hub/checkpoints/resnet18-5c106cde.pth


In [ ]:
resnet18

ResNet(
  (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
  (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (layer1): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
  

In [ ]:
resnet18_pt

ResNet(
  (conv1): Conv2d(3, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)
  (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (relu): ReLU(inplace=True)
  (maxpool): MaxPool2d(kernel_size=3, stride=2, padding=1, dilation=1, ceil_mode=False)
  (layer1): Sequential(
    (0): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
      (conv2): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn2): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
    )
    (1): BasicBlock(
      (conv1): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), bias=False)
      (bn1): BatchNorm2d(64, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
      (relu): ReLU(inplace=True)
  

In [ ]:
alexnet = models.alexnet()
alexnet_pt = models.alexnet(pretrained=True)

Downloading: "https://download.pytorch.org/models/alexnet-owt-4df8aa71.pth" to /root/.cache/torch/hub/checkpoints/alexnet-owt-4df8aa71.pth


In [ ]:
alexnet

AlexNet(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(11, 11), stride=(4, 4), padding=(2, 2))
    (1): ReLU(inplace=True)
    (2): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
    (3): Conv2d(64, 192, kernel_size=(5, 5), stride=(1, 1), padding=(2, 2))
    (4): ReLU(inplace=True)
    (5): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
    (6): Conv2d(192, 384, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (7): ReLU(inplace=True)
    (8): Conv2d(384, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (9): ReLU(inplace=True)
    (10): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): ReLU(inplace=True)
    (12): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(6, 6))
  (classifier): Sequential(
    (0): Dropout(p=0.5, inplace=False)
    (1): Linear(in_features=9216, out_features=4096, bias=True)
 

In [ ]:
alexnet_pt

AlexNet(
  (features): Sequential(
    (0): Conv2d(3, 64, kernel_size=(11, 11), stride=(4, 4), padding=(2, 2))
    (1): ReLU(inplace=True)
    (2): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
    (3): Conv2d(64, 192, kernel_size=(5, 5), stride=(1, 1), padding=(2, 2))
    (4): ReLU(inplace=True)
    (5): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
    (6): Conv2d(192, 384, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (7): ReLU(inplace=True)
    (8): Conv2d(384, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (9): ReLU(inplace=True)
    (10): Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
    (11): ReLU(inplace=True)
    (12): MaxPool2d(kernel_size=3, stride=2, padding=0, dilation=1, ceil_mode=False)
  )
  (avgpool): AdaptiveAvgPool2d(output_size=(6, 6))
  (classifier): Sequential(
    (0): Dropout(p=0.5, inplace=False)
    (1): Linear(in_features=9216, out_features=4096, bias=True)
 

In [ ]:
sample_input=torch.rand(1,3,224,224)

In [ ]:
print(resnet18(sample_input))
print(resnet18_pt(sample_input))
print(alexnet(sample_input))
print(alexnet_pt(sample_input))

tensor([[-3.7480e-01,  1.7617e-01, -5.9083e-01, -4.7130e-01, -6.8953e-01,
          9.0782e-02, -4.4005e-02, -4.7955e-01, -2.5937e-01,  2.0701e-01,
          1.7559e-01, -1.0990e+00,  6.9854e-01, -1.7944e+00, -9.0110e-01,
         -9.4805e-01, -9.3713e-01, -5.1444e-01,  1.8877e-01, -7.2581e-02,
         -3.5521e-01, -3.7642e-01,  3.6675e-01,  4.4044e-01,  5.4703e-03,
         -5.7668e-01,  5.8538e-01, -1.2001e+00, -2.2794e-01,  5.8323e-01,
         -1.5892e-01, -2.2846e-02, -7.6490e-01, -9.5614e-01,  8.2606e-01,
          1.9579e-01, -4.0407e-01, -2.6577e-01,  3.5535e-01, -2.9903e-01,
         -1.1595e+00, -1.5173e-01, -1.8430e-01,  1.4570e-01, -3.2693e-01,
          6.6518e-01,  7.7033e-01,  1.1847e-01,  3.7505e-01, -5.9252e-01,
         -4.1773e-03,  4.5659e-01,  2.3347e-01, -1.0798e-01, -4.1171e-01,
          8.8393e-01, -3.8764e-01,  2.4220e-01, -4.6403e-01,  9.5299e-01,
          3.5307e-02,  3.0320e-01,  3.9352e-01, -3.9524e-01,  1.9584e-01,
          2.4106e-02,  1.4316e-01,  6.

We'll talk about this soon (in Transfer Learning)
- See this tutorial: [Fine tuning Torchvision Models](https://pytorch.org/tutorials/beginner/finetuning_torchvision_models_tutorial.html

# **Regularization** for Deep Learning (Part two): 

- Ensemble Methods: Bagging 
- Dropout
- Adversarial Training: Example Generation and Robustness (Two detailed examples - I'll refer you to them) : [Tutorial1](https://pytorch.org/tutorials/beginner/fgsm_tutorial.html) - [Tutorial2](https://adversarial-ml-tutorial.org/introduction/)  

- Physics Informed Neural Network (A bright exmple of regularization technique) : See this [PINNs TF Implementation](https://github.com/pierremtb/PINNs-TF2.0)


**Ensemble Methods**: Bagging as well as other methods

* Building a custom (ensemble) model in Pytorch. Take 
 a look at this [dicussion](https://discuss.pytorch.org/t/custom-ensemble-approach/52024)
* [Ensemble-Pytorch](https://github.com/AaronX121/Ensemble-Pytorch): Implementation of scikit-learn like ensemble methods in Pytorch
* Also, The [sklearn.ensemble](https://scikit-learn.org/stable/modules/ensemble.html) module provides extensive ensembles based on available models in [sklearn](https://scikit-learn.org/stable/index.html)

# Dropout

First a bench mark model: 


In [ ]:
# Modified based on https://github.com/pytorch/examples/tree/master/mnist

from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
       # self.dropout1 = nn.Dropout(0.25)
       # self.dropout2 = nn.Dropout(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
       # x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
       # x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
      loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
            epoch, batch_idx * len(data), len(train_loader.dataset),
            100. * batch_idx / len(train_loader), loss.item()))


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))



In [ ]:
    torch.manual_seed(112)
    device =torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")

In [ ]:
    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset_train = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset_test = datasets.MNIST('../data', train=False,
                       transform=transform)

Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw



Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw
Processing...
Done!


/usr/local/lib/python3.6/dist-packages/torchvision/datasets/mnist.py:469: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


In [ ]:
dataset_train, dataset_test

(Dataset MNIST
     Number of datapoints: 60000
     Root location: ../data
     Split: Train
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ), Dataset MNIST
     Number of datapoints: 10000
     Root location: ../data
     Split: Test
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ))

In [ ]:
torch.manual_seed(123)
transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)

train_loader = torch.utils.data.DataLoader(dataset_train, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)
test_loader = torch.utils.data.DataLoader(dataset_test, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)

model = Net().to(device)
optimizer = optim.Adadelta(model.parameters(), lr=0.01)

scheduler = StepLR(optimizer, step_size=5, gamma=0.1 )
for epoch in range(1, 20 + 1):
  train(10, model, device, train_loader, optimizer, epoch)
  test(model, device, test_loader)
  scheduler.step()

torch.save(model.state_dict(), "mnist_1.pt")


Train Epoch: 1 [0/60000 (0%)]	Loss: 2.307113
Train Epoch: 1 [2560/60000 (4%)]	Loss: 2.246330
Train Epoch: 1 [5120/60000 (9%)]	Loss: 2.152776
Train Epoch: 1 [7680/60000 (13%)]	Loss: 2.053525
Train Epoch: 1 [10240/60000 (17%)]	Loss: 1.883935
Train Epoch: 1 [12800/60000 (21%)]	Loss: 1.724492
Train Epoch: 1 [15360/60000 (26%)]	Loss: 1.479796
Train Epoch: 1 [17920/60000 (30%)]	Loss: 1.235827
Train Epoch: 1 [20480/60000 (34%)]	Loss: 1.034860
Train Epoch: 1 [23040/60000 (38%)]	Loss: 0.853716
Train Epoch: 1 [25600/60000 (43%)]	Loss: 0.772773
Train Epoch: 1 [28160/60000 (47%)]	Loss: 0.717264
Train Epoch: 1 [30720/60000 (51%)]	Loss: 0.640164
Train Epoch: 1 [33280/60000 (55%)]	Loss: 0.587384
Train Epoch: 1 [35840/60000 (60%)]	Loss: 0.539963
Train Epoch: 1 [38400/60000 (64%)]	Loss: 0.491984
Train Epoch: 1 [40960/60000 (68%)]	Loss: 0.530113
Train Epoch: 1 [43520/60000 (72%)]	Loss: 0.548230
Train Epoch: 1 [46080/60000 (77%)]	Loss: 0.513601
Train Epoch: 1 [48640/60000 (81%)]	Loss: 0.443843
Train Epoc

Now let's start:

 **Torch.dropout**: https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html

In [ ]:
dropout_layer = nn.Dropout(1)
input = torch.randn(2, 20)
output = dropout_layer(input)
print(input)
print(output)

tensor([[-0.2393, -0.8378,  0.8559, -1.3261,  0.0330,  1.2033,  0.4886, -0.2248,
          2.1273,  0.3787, -0.8657,  1.5655, -0.9437, -1.9373,  0.1816, -0.5516,
         -1.7298,  1.1813, -0.1465, -1.0134],
        [-0.0535,  1.7472, -0.8287, -0.6501, -0.1857,  1.2836,  0.6029, -0.2008,
         -2.5054,  1.4367,  0.9614, -0.8520, -1.1531, -2.1133, -0.2215,  1.1383,
          0.1116, -1.4435, -0.9796, -0.3777]])
tensor([[-0., -0., 0., -0., 0., 0., 0., -0., 0., 0., -0., 0., -0., -0., 0., -0., -0., 0., -0., -0.],
        [-0., 0., -0., -0., -0., 0., 0., -0., -0., 0., 0., -0., -0., -0., -0., 0., 0., -0., -0., -0.]])


In [ ]:
dropout_layer = nn.Dropout(0.25)
input = torch.randn(2, 20)
output = dropout_layer(input)
print(input)
print(output)

tensor([[ 2.0883, -0.3782, -0.1384,  0.4667,  0.7491,  1.3325, -0.4197,  1.2035,
         -0.1668,  1.5479, -0.8813,  0.7535,  0.3926, -0.2745, -0.6635,  1.1863,
          1.7974, -0.2479, -0.3005,  1.4417],
        [-1.0637, -0.1633,  1.2450, -0.8118,  0.0041, -0.8165,  0.3768, -1.0938,
         -1.5452, -1.2712, -1.1258, -1.6945, -0.5335, -0.1772,  0.3597, -1.6240,
          0.1669,  0.0416, -1.1723, -0.5961]])
tensor([[ 2.7844, -0.5042, -0.0000,  0.6223,  0.9988,  1.7767, -0.5596,  0.0000,
         -0.0000,  0.0000, -1.1750,  1.0047,  0.5235, -0.3660, -0.8846,  0.0000,
          2.3965, -0.0000, -0.4007,  1.9222],
        [-1.4182, -0.2178,  0.0000, -0.0000,  0.0055, -1.0887,  0.5024, -1.4584,
         -2.0603, -1.6949, -1.5011, -2.2593, -0.7113, -0.2363,  0.4796, -2.1653,
          0.2226,  0.0555, -1.5631, -0.7948]])


**First Experiment**

In [ ]:
# Modified based on https://github.com/pytorch/examples/tree/master/mnist

from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.25)
        self.dropout2 = nn.Dropout(0.5)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
            epoch, batch_idx * len(data), len(train_loader.dataset),
            100. * batch_idx / len(train_loader), loss.item()))


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))



In [ ]:
    torch.manual_seed(112)
    device =torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")

In [ ]:
    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset_train = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset_test = datasets.MNIST('../data', train=False,
                       transform=transform)

Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw



Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw
Processing...
Done!


/usr/local/lib/python3.6/dist-packages/torchvision/datasets/mnist.py:480: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


In [ ]:
dataset_train, dataset_test

(Dataset MNIST
     Number of datapoints: 60000
     Root location: ../data
     Split: Train
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ), Dataset MNIST
     Number of datapoints: 10000
     Root location: ../data
     Split: Test
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ))

In [ ]:
torch.manual_seed(123)
transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)

train_loader = torch.utils.data.DataLoader(dataset_train, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)
test_loader = torch.utils.data.DataLoader(dataset_test, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)

model = Net().to(device)
optimizer = optim.Adadelta(model.parameters(), lr=0.01)

scheduler = StepLR(optimizer, step_size=5, gamma=0.1 )
for epoch in range(1, 20 + 1):
  train(10, model, device, train_loader, optimizer, epoch)
  test(model, device, test_loader)
  scheduler.step()

torch.save(model.state_dict(), "mnist_51.pt")



Train Epoch: 1 [0/60000 (0%)]	Loss: 2.290775
Train Epoch: 1 [2560/60000 (4%)]	Loss: 2.263582
Train Epoch: 1 [5120/60000 (9%)]	Loss: 2.178796
Train Epoch: 1 [7680/60000 (13%)]	Loss: 2.112189
Train Epoch: 1 [10240/60000 (17%)]	Loss: 2.012206
Train Epoch: 1 [12800/60000 (21%)]	Loss: 1.875768
Train Epoch: 1 [15360/60000 (26%)]	Loss: 1.732704
Train Epoch: 1 [17920/60000 (30%)]	Loss: 1.570873
Train Epoch: 1 [20480/60000 (34%)]	Loss: 1.368344
Train Epoch: 1 [23040/60000 (38%)]	Loss: 1.252594
Train Epoch: 1 [25600/60000 (43%)]	Loss: 1.100150
Train Epoch: 1 [28160/60000 (47%)]	Loss: 1.010099
Train Epoch: 1 [30720/60000 (51%)]	Loss: 0.966332
Train Epoch: 1 [33280/60000 (55%)]	Loss: 0.941042
Train Epoch: 1 [35840/60000 (60%)]	Loss: 0.841713
Train Epoch: 1 [38400/60000 (64%)]	Loss: 0.837612
Train Epoch: 1 [40960/60000 (68%)]	Loss: 0.686441
Train Epoch: 1 [43520/60000 (72%)]	Loss: 0.746758
Train Epoch: 1 [46080/60000 (77%)]	Loss: 0.710856
Train Epoch: 1 [48640/60000 (81%)]	Loss: 0.590436
Train Epo

**Second Experiment**

In [ ]:
# Modified based on https://github.com/pytorch/examples/tree/master/mnist

from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.1)
        self.dropout2 = nn.Dropout(0.1)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
            epoch, batch_idx * len(data), len(train_loader.dataset),
            100. * batch_idx / len(train_loader), loss.item()))


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))



In [ ]:
    torch.manual_seed(112)
    device =torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")

In [ ]:
    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset_train = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset_test = datasets.MNIST('../data', train=False,
                       transform=transform)

In [ ]:
torch.manual_seed(123)
transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)

train_loader = torch.utils.data.DataLoader(dataset_train, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)
test_loader = torch.utils.data.DataLoader(dataset_test, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)

model = Net().to(device)
optimizer = optim.Adadelta(model.parameters(), lr=0.01)

scheduler = StepLR(optimizer, step_size=5, gamma=0.1 )
for epoch in range(1, 20 + 1):
  train(10, model, device, train_loader, optimizer, epoch)
  test(model, device, test_loader)
  scheduler.step()

torch.save(model.state_dict(), "mnist_51.pt")


Train Epoch: 1 [0/60000 (0%)]	Loss: 2.290485
Train Epoch: 1 [2560/60000 (4%)]	Loss: 2.242163
Train Epoch: 1 [5120/60000 (9%)]	Loss: 2.153081
Train Epoch: 1 [7680/60000 (13%)]	Loss: 2.053765
Train Epoch: 1 [10240/60000 (17%)]	Loss: 1.915162
Train Epoch: 1 [12800/60000 (21%)]	Loss: 1.728943
Train Epoch: 1 [15360/60000 (26%)]	Loss: 1.546496
Train Epoch: 1 [17920/60000 (30%)]	Loss: 1.316280
Train Epoch: 1 [20480/60000 (34%)]	Loss: 1.071331
Train Epoch: 1 [23040/60000 (38%)]	Loss: 0.965430
Train Epoch: 1 [25600/60000 (43%)]	Loss: 0.813209
Train Epoch: 1 [28160/60000 (47%)]	Loss: 0.739535
Train Epoch: 1 [30720/60000 (51%)]	Loss: 0.623678
Train Epoch: 1 [33280/60000 (55%)]	Loss: 0.664210
Train Epoch: 1 [35840/60000 (60%)]	Loss: 0.566425
Train Epoch: 1 [38400/60000 (64%)]	Loss: 0.585027
Train Epoch: 1 [40960/60000 (68%)]	Loss: 0.465509
Train Epoch: 1 [43520/60000 (72%)]	Loss: 0.512838
Train Epoch: 1 [46080/60000 (77%)]	Loss: 0.518450
Train Epoch: 1 [48640/60000 (81%)]	Loss: 0.403659
Train Epoc

# **Optimization** for Training Deep Models: Different Optimizers and Batch Normalization

* https://pytorch.org/docs/stable/optim.html
* https://github.com/jettify/pytorch-optimizer
* https://www.deeplearningwizard.com/deep_learning/boosting_models_pytorch/optimizers/
* See this: https://github.com/ritchieng/deep-learning-wizard/tree/master/docs/deep_learning/boosting_models_pytorch

# Different Optimizers 
* Browed from: https://github.com/ritchieng/deep-learning-wizard/tree/master/docs/deep_learning/boosting_models_pytorch

### Model: 1 Hidden Layer Feedforward Neural Network (ReLU Activation)
<img src="./images/nn1.png" alt="deeplearningwizard" style="width: 900px;"/>

### Steps
- Step 1: Load Dataset
- Step 2: Make Dataset Iterable
- Step 3: Create Model Class
- Step 4: Instantiate Model Class
- Step 5: Instantiate Loss Class
- **Step 6: Instantiate Optimizer Class**
- Step 7: Train Model

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)


'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
learning_rate = 0.1

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Extracting ./data/MNIST/raw/train-images-idx3-ubyte.gz to ./data/MNIST/raw


Extracting ./data/MNIST/raw/train-labels-idx1-ubyte.gz to ./data/MNIST/raw



Extracting ./data/MNIST/raw/t10k-images-idx3-ubyte.gz to ./data/MNIST/raw


Extracting ./data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ./data/MNIST/raw
Processing...
Done!


/usr/local/lib/python3.6/dist-packages/torchvision/datasets/mnist.py:480: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)



Iteration: 500. Loss: 0.26971518993377686. Accuracy: 91.51000213623047
Iteration: 1000. Loss: 0.1802702695131302. Accuracy: 92.87999725341797
Iteration: 1500. Loss: 0.1705324351787567. Accuracy: 94.13999938964844
Iteration: 2000. Loss: 0.2071806788444519. Accuracy: 94.68000030517578
Iteration: 2500. Loss: 0.15649180114269257. Accuracy: 95.16000366210938
Iteration: 3000. Loss: 0.13750530779361725. Accuracy: 95.83999633789062


### Non-Technical Process 
1. Convert inputs/labels to variables
2. Clear gradient buffers
3. Get output given inputs 
4. Get loss by comparing with labels
5. Get gradients w.r.t. parameters
6. **Update parameters using gradients**
    - `parameters = parameters - learning_rate * parameters_gradients`
7. REPEAT

### Why is it called Gradient Descent? 
- Use gradients (error signals) $\rightarrow$ update parameters to minimize our loss (descent) $\rightarrow$ better predictive accuracy


### Mathematical Interpretation of Gradient Descent
- Model's parameters: $\theta \in ℝ^d$
- Loss function: $J(\theta)$
- Gradient w.r.t. parameters: $ \nabla J(\theta)$
- Learning rate: $\eta$
- Batch Gradient descent: $\theta = \theta - \eta \cdot  \nabla J(\theta)$


## Optimization Algorithm 1: Batch Gradient Descent
- What we've covered so far: batch gradient descent
    - $\theta = \theta - \eta \cdot  \nabla J(\theta)$
- Characteristics
    - Compute the gradient of the lost function w.r.t. parameters for the entire training data, $\nabla J(\theta)$ 
    - Use this to update our parameters at every iteration
- Problems
    - Unable to fit whole datasets in memory 
    - Computationally slow as we attempt to compute a large Jacobian matrix $\rightarrow$ first order derivative, $\nabla J(\theta)$
- Conceptually easy to understand $\rightarrow$ rarely used

## Optimization Algorithm 2: Stochastic Gradient Descent 
- Modification of batch gradient descent
    - $\theta = \theta - \eta \cdot  \nabla J(\theta, x^{i}, y^{i})$
- Characteristics
    - Compute the gradient of the lost function w.r.t. parameters for the **one set of training sample (1 input and 1 label)**, $\nabla J(\theta, x^{i}, y^{i})$
    - Use this to update our parameters at every iteration
- Benefits
    - Able to fit large datasets
    - Computationally faster $\rightarrow$ instead gradients w.r.t to the whole training data, we get the gradients w.r.t. training sample
- Problems
    - Updating very frequently $\rightarrow$ huge variance in parameter updates $\rightarrow$ may overshoot local minima 
        - Can be solved by carefully decaying your learning rate $\rightarrow$ take smaller steps in incorporating gradients to improve the parameters

## Optimization Algorithm 3: Mini-batch Gradient Descent
- Combination of batch gradient descent & stochastic gradient descent
    - $\theta = \theta - \eta \cdot  \nabla J(\theta, x^{i: i+n}, y^{i:i+n})$
- Characteristics
    - Compute the gradient of the lost function w.r.t. parameters for **n sets of training sample (n input and n label)**, $\nabla J(\theta, x^{i: i+n}, y^{i:i+n})$
    - Use this to update our parameters at every iteration
- Benefits
    - Able to fit large datasets
    - Computationally faster $\rightarrow$ instead gradients w.r.t to the whole training data, we get the gradients w.r.t. training sample
    - Lower variance of parameter updates
- This is often called SGD in deep learning frameworks .__.  

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim)
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
learning_rate = 0.1

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate)

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28).requires_grad_()
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.26971518993377686. Accuracy: 91.51000213623047
Iteration: 1000. Loss: 0.1802702695131302. Accuracy: 92.87999725341797
Iteration: 1500. Loss: 0.1705324351787567. Accuracy: 94.13999938964844
Iteration: 2000. Loss: 0.2071806788444519. Accuracy: 94.68000030517578
Iteration: 2500. Loss: 0.15649180114269257. Accuracy: 95.16000366210938
Iteration: 3000. Loss: 0.13750530779361725. Accuracy: 95.83999633789062


## Optimization Algorithm 4: SGD Momentum
- Modification of SGD
    - $v_t = \gamma v_{t-1} + \eta \cdot  \nabla J(\theta, x^{i: i+n}, y^{i:i+n})$
    - $\theta = \theta - v_t$
- Characteristics
    - Compute the gradient of the lost function w.r.t. parameters for **n sets of training sample (n input and n label)**, $\nabla J(\theta, x^{i: i+n}, y^{i:i+n})$
    - Use this to add to the previous update vector $v_{t-1}$
    - Momentum, usually set to $\gamma = 0.9$
    - Parameters updated with update vector, $v_t$ that incorporates previous update vector
        - $\gamma v_{t}$ increases if gradient same sign/direction as $v_{t-1}$ 
            - Gives SGD the push when it is going in the right direction (minimizing loss)
            - Accelerated convergence
        - $\gamma v_{t}$ decreases if gradient different sign/direction as $v_{t-1}$
            - Dampens SGD when it is going in a different direction
            - Lower variation in loss minimization
- Problems
    - It might go the wrong direction (higher loss) $\rightarrow$ continue to be accelerated to the wrong direction (higher loss) 


In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
learning_rate = 0.1

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0.9)

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.11467370390892029. Accuracy: 96.19999694824219
Iteration: 1000. Loss: 0.056595753878355026. Accuracy: 96.55000305175781
Iteration: 1500. Loss: 0.15522664785385132. Accuracy: 97.16999816894531
Iteration: 2000. Loss: 0.046876028180122375. Accuracy: 97.30999755859375
Iteration: 2500. Loss: 0.058556076139211655. Accuracy: 97.16999816894531
Iteration: 3000. Loss: 0.08808634430170059. Accuracy: 97.52999877929688


## Optimization Algorithm 4: SGD Nesterov
- Modification of SGD Momentum 
    - $v_t = \gamma v_{t-1} + \eta \cdot  \nabla J(\theta - \gamma v_{t-1}, x^{i: i+n}, y^{i:i+n})$
    - $\theta = \theta - v_t$
- Characteristics
    - Compute the gradient of the lost function w.r.t. **future approximate parameters** for **n sets of training sample (n input and n label)**, $\nabla J(\theta - \gamma v_{t-1}, x^{i: i+n}, y^{i:i+n})$
        - Use this to add to the previous update vector $v_{t-1}$
        - Momentum, usually set to $\gamma = 0.9$
    - Gradients w.r.t. future approximate parameters $\rightarrow$ sense of where we will be $\rightarrow$ anticipate if we are going in the wrong direction in the next step $\rightarrow$ slow down accordingly

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
learning_rate = 0.1

optimizer = torch.optim.SGD(model.parameters(), lr=learning_rate, momentum=0.9, nesterov=True)

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.15253432095050812. Accuracy: 96.38999938964844
Iteration: 1000. Loss: 0.055656854063272476. Accuracy: 96.44000244140625
Iteration: 1500. Loss: 0.13686475157737732. Accuracy: 97.18000030517578
Iteration: 2000. Loss: 0.07397746294736862. Accuracy: 97.45999908447266
Iteration: 2500. Loss: 0.07314040511846542. Accuracy: 97.37999725341797
Iteration: 3000. Loss: 0.06474008411169052. Accuracy: 97.44999694824219


## Optimization Algorithm 4: Adam
- Adaptive Learning Rates
    - $m_t = \beta_1 m_{t-1} + (1 - \beta_1)g_t$
        - Keeping track of decaying gradient
        - Estimate of the mean of gradients
    - $v_t = \beta_2 v_{t-1} + (1 - \beta_2)g_t^2$
        - Keeping track of decaying squared gradient 
        - Estimate of the variance of gradients
    - When $m_t, v_t$ initializes as 0, $m_t, v_t \rightarrow 0$ initially when decay rates small, $\beta_1, \beta_2 \rightarrow 1$  
        - Need to correct this with:
        - $ \hat m_t = \frac{m_t}{1- \beta_1}$
        - $ \hat v_t = \frac{v_t}{1- \beta_2}$
    - $\theta_{t+1} = \theta_t - \frac{\eta}{\sqrt{\hat v_t} + \epsilon}\hat m_t$
    - Default recommended values
        - $\beta_1 = 0.9$
        - $\beta_2 = 0.999$
        - $\epsilon = 10^{-8}$
- Instead of learning rate $\rightarrow$ equations account for estimates of mean/variance of gradients to determine the next learning rate


In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
# learning_rate = 0.001

optimizer = torch.optim.Adam(model.parameters())

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.23362819850444794. Accuracy: 93.3499984741211
Iteration: 1000. Loss: 0.11055783182382584. Accuracy: 94.8499984741211
Iteration: 1500. Loss: 0.14103564620018005. Accuracy: 95.95999908447266
Iteration: 2000. Loss: 0.107691690325737. Accuracy: 96.55000305175781
Iteration: 2500. Loss: 0.0956580713391304. Accuracy: 96.56999969482422
Iteration: 3000. Loss: 0.07698246836662292. Accuracy: 97.16999816894531


## Other Adaptive Algorithms
- Other adaptive algorithms (like Adam, adapting learning rates)
    - Adagrad
    - Adadelta
    - Adamax
    - RMSProp

## Optimization Algorithm 5: Adagrad

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
# learning_rate = 0.001

optimizer = torch.optim.Adagrad(model.parameters())

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.24082742631435394. Accuracy: 93.0
Iteration: 1000. Loss: 0.15385352075099945. Accuracy: 93.9800033569336
Iteration: 1500. Loss: 0.1544286012649536. Accuracy: 94.80999755859375
Iteration: 2000. Loss: 0.17311468720436096. Accuracy: 95.12000274658203
Iteration: 2500. Loss: 0.19428884983062744. Accuracy: 95.37999725341797
Iteration: 3000. Loss: 0.14685769379138947. Accuracy: 95.66999816894531


## Optimization Algorithm 6: Adadelta

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets
from torch.autograd import Variable

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
# learning_rate = 0.001

optimizer = torch.optim.Adadelta(model.parameters())

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variablee
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.16266268491744995. Accuracy: 94.79000091552734
Iteration: 1000. Loss: 0.06723383069038391. Accuracy: 95.98999786376953
Iteration: 1500. Loss: 0.11248779296875. Accuracy: 96.94999694824219
Iteration: 2000. Loss: 0.06128310784697533. Accuracy: 97.18000030517578
Iteration: 2500. Loss: 0.05549431964755058. Accuracy: 97.25
Iteration: 3000. Loss: 0.04269982874393463. Accuracy: 97.37999725341797


## Optimization Algorithm 6: Adamax

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
# learning_rate = 0.001

optimizer = torch.optim.Adamax(model.parameters())

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28)
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.24909545481204987. Accuracy: 92.19000244140625
Iteration: 1000. Loss: 0.1435246467590332. Accuracy: 93.61000061035156
Iteration: 1500. Loss: 0.15786641836166382. Accuracy: 94.9800033569336
Iteration: 2000. Loss: 0.14738872647285461. Accuracy: 95.48999786376953
Iteration: 2500. Loss: 0.14099086821079254. Accuracy: 95.81999969482422
Iteration: 3000. Loss: 0.11718937754631042. Accuracy: 96.55000305175781


## Optimization Algorithm 7: RMSProp

In [ ]:
import torch
import torch.nn as nn
import torchvision.transforms as transforms
import torchvision.datasets as dsets

# Set seed
torch.manual_seed(0)

'''
STEP 1: LOADING DATASET
'''

train_dataset = dsets.MNIST(root='./data', 
                            train=True, 
                            transform=transforms.ToTensor(),
                            download=True)

test_dataset = dsets.MNIST(root='./data', 
                           train=False, 
                           transform=transforms.ToTensor())

'''
STEP 2: MAKING DATASET ITERABLE
'''

batch_size = 100
n_iters = 3000
num_epochs = n_iters / (len(train_dataset) / batch_size)
num_epochs = int(num_epochs)

train_loader = torch.utils.data.DataLoader(dataset=train_dataset, 
                                           batch_size=batch_size, 
                                           shuffle=True)

test_loader = torch.utils.data.DataLoader(dataset=test_dataset, 
                                          batch_size=batch_size, 
                                          shuffle=False)

'''
STEP 3: CREATE MODEL CLASS
'''
class FeedforwardNeuralNetModel(nn.Module):
    def __init__(self, input_dim, hidden_dim, output_dim):
        super(FeedforwardNeuralNetModel, self).__init__()
        # Linear function
        self.fc1 = nn.Linear(input_dim, hidden_dim) 
        # Non-linearity
        self.relu = nn.ReLU()
        # Linear function (readout)
        self.fc2 = nn.Linear(hidden_dim, output_dim)  
    
    def forward(self, x):
        # Linear function
        out = self.fc1(x)
        # Non-linearity
        out = self.relu(out)
        # Linear function (readout)
        out = self.fc2(out)
        return out
'''
STEP 4: INSTANTIATE MODEL CLASS
'''
input_dim = 28*28
hidden_dim = 100
output_dim = 10

model = FeedforwardNeuralNetModel(input_dim, hidden_dim, output_dim)

'''
STEP 5: INSTANTIATE LOSS CLASS
'''
criterion = nn.CrossEntropyLoss()


'''
STEP 6: INSTANTIATE OPTIMIZER CLASS
'''
# learning_rate = 0.001

optimizer = torch.optim.RMSprop(model.parameters())

'''
STEP 7: TRAIN THE MODEL
'''
iter = 0
for epoch in range(num_epochs):
    for i, (images, labels) in enumerate(train_loader):
        # Load images as Variable
        images = images.view(-1, 28*28).requires_grad_()
        
        # Clear gradients w.r.t. parameters
        optimizer.zero_grad()
        
        # Forward pass to get output/logits
        outputs = model(images)
        
        # Calculate Loss: softmax --> cross entropy loss
        loss = criterion(outputs, labels)
        
        # Getting gradients w.r.t. parameters
        loss.backward()
        
        # Updating parameters
        optimizer.step()
        
        iter += 1
        
        if iter % 500 == 0:
            # Calculate Accuracy         
            correct = 0
            total = 0
            # Iterate through test dataset
            for images, labels in test_loader:
                # Load images to a Torch Variable
                images = images.view(-1, 28*28).requires_grad_()
                
                # Forward pass only to get logits/output
                outputs = model(images)
                
                # Get predictions from the maximum value
                _, predicted = torch.max(outputs.data, 1)
                
                # Total number of labels
                total += labels.size(0)
                
                # Total correct predictions
                correct += (predicted == labels).sum()
            
            accuracy = 100 * correct / total
            
            # Print Loss
            print('Iteration: {}. Loss: {}. Accuracy: {}'.format(iter, loss.item(), accuracy))

Iteration: 500. Loss: 0.1688031703233719. Accuracy: 95.13999938964844
Iteration: 1000. Loss: 0.035940684378147125. Accuracy: 95.55999755859375
Iteration: 1500. Loss: 0.13070625066757202. Accuracy: 96.19000244140625
Iteration: 2000. Loss: 0.12187916785478592. Accuracy: 95.56999969482422
Iteration: 2500. Loss: 0.22075626254081726. Accuracy: 96.55999755859375
Iteration: 3000. Loss: 0.06314142048358917. Accuracy: 96.76000213623047


## Summary of Optimization Algorithms Performance
- SGD: 95.78%
- **SGD Momentum: 97.69%**
- **SGD Nesterov: 97.58%**
- **Adam: 97.20%**
- Adagrad: 95.51%
- **Adadelta: 97.45%**
- Adamax: 96.58%
- **RMSProp: 97.1%**

## Simple Suggestions
- Momentum/Nesterov
    - Powerful if we control the learning rate schedule
- Adam
    - Lazy to control the learning rate schedule

## Summary
- **Recap of 7 step process**
    - Step 1: Load Dataset
    - Step 2: Make Dataset Iterable
    - Step 3: Create Model Class
    - Step 4: Instantiate Model Class
    - Step 5: Instantiate Loss Class
    - **Step 6: Instantiate Optimizer Class**
    - Step 7: Train Model
- **Step 6**
    - Update parameters using gradients
    - `parameters = parameters - learning_rate * parameters_gradients`
- **Gradient descent**
    - Using gradients (error signals from loss class) to update parameters
- **Mathematical** interpretation: $\theta = \theta - \eta \cdot  \nabla J(\theta)$
- **Optimisation Algorithms**
    - Batch gradient descent
    - Stochastic gradient descent
    - Mini-batch gradient descent (SGD)
    - SGD + Momentum
    - SGD + Nesterov
    - Adam
    - Other adaptive algorithms: adagrad, adamax, adadelta, RMSProp
- **Recommendations**
    - SGD+M
    - SGD+N
    - Adam

See Tune for hyperparameter optimization: https://docs.ray.io/en/master/tune/

And: https://pytorch.org/tutorials/beginner/hyperparameter_tuning_tutorial.html

This is a must (super informative): https://playground.tensorflow.org/ 

# Now let's work on our benchmark model: This time with Adam optimizer

In [ ]:
# Modified based on https://github.com/pytorch/examples/tree/master/mnist

from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.dropout1 = nn.Dropout(0.1)
        self.dropout2 = nn.Dropout(0.1)
        self.fc1 = nn.Linear(9216, 128)
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
        x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.dropout2(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
            epoch, batch_idx * len(data), len(train_loader.dataset),
            100. * batch_idx / len(train_loader), loss.item()))


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))



In [ ]:
    torch.manual_seed(112)
    device =torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")

In [ ]:
    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset_train = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset_test = datasets.MNIST('../data', train=False,
                       transform=transform)

Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw



Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw
Processing...
Done!


/usr/local/lib/python3.6/dist-packages/torchvision/datasets/mnist.py:480: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


In [ ]:
dataset_train, dataset_test

(Dataset MNIST
     Number of datapoints: 60000
     Root location: ../data
     Split: Train
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ), Dataset MNIST
     Number of datapoints: 10000
     Root location: ../data
     Split: Test
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ))

In [ ]:
torch.manual_seed(123)
transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)

train_loader = torch.utils.data.DataLoader(dataset_train, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)
test_loader = torch.utils.data.DataLoader(dataset_test, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)

model = Net().to(device)
#optimizer = optim.Adadelta(model.parameters(), lr=0.01)
optimizer=optim.Adam(model.parameters())
# see the Adam optimizer precisely: https://pytorch.org/docs/stable/_modules/torch/optim/adam.html#Adam

#scheduler = StepLR(optimizer, step_size=5, gamma=0.1 )
for epoch in range(1, 20 + 1):
  train(10, model, device, train_loader, optimizer, epoch)
  test(model, device, test_loader)
 # scheduler.step()

torch.save(model.state_dict(), "mnist_51.pt")


Train Epoch: 1 [0/60000 (0%)]	Loss: 2.290485
Train Epoch: 1 [2560/60000 (4%)]	Loss: 0.499287
Train Epoch: 1 [5120/60000 (9%)]	Loss: 0.340078
Train Epoch: 1 [7680/60000 (13%)]	Loss: 0.307996
Train Epoch: 1 [10240/60000 (17%)]	Loss: 0.230379
Train Epoch: 1 [12800/60000 (21%)]	Loss: 0.192829
Train Epoch: 1 [15360/60000 (26%)]	Loss: 0.120983
Train Epoch: 1 [17920/60000 (30%)]	Loss: 0.170480
Train Epoch: 1 [20480/60000 (34%)]	Loss: 0.116984
Train Epoch: 1 [23040/60000 (38%)]	Loss: 0.088619
Train Epoch: 1 [25600/60000 (43%)]	Loss: 0.084167
Train Epoch: 1 [28160/60000 (47%)]	Loss: 0.113863
Train Epoch: 1 [30720/60000 (51%)]	Loss: 0.074733
Train Epoch: 1 [33280/60000 (55%)]	Loss: 0.117818
Train Epoch: 1 [35840/60000 (60%)]	Loss: 0.105022
Train Epoch: 1 [38400/60000 (64%)]	Loss: 0.100068
Train Epoch: 1 [40960/60000 (68%)]	Loss: 0.082639
Train Epoch: 1 [43520/60000 (72%)]	Loss: 0.073992
Train Epoch: 1 [46080/60000 (77%)]	Loss: 0.050055
Train Epoch: 1 [48640/60000 (81%)]	Loss: 0.045510
Train Epoc

# **Batch Normalization**

* BatchNorm2d: https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm2d.html
* BatchNorm1d: https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html

In [ ]:
# Without Learnable Parameters
BatchNorm1d_layer = nn.BatchNorm1d(10, affine=False)
# With Learnable Parameters
BatchNorm1d_layer = nn.BatchNorm1d(10)
input = torch.randn(20, 10)
output = BatchNorm1d_layer(input)
print(input)
print(output)

tensor([[ 6.5536e-02,  1.2142e+00,  4.9243e-01,  1.2853e+00,  4.6950e-01,
          2.1183e+00,  2.4904e+00, -1.1194e+00,  5.9126e-01,  2.0532e-01],
        [ 4.4115e-01,  3.7207e-01,  1.6566e-01, -2.0475e+00,  1.7869e+00,
          3.5882e-01,  1.5658e+00, -1.7263e-01,  2.6234e-01, -1.1125e-01],
        [ 1.3818e+00, -9.7413e-02, -3.8611e-01, -4.2693e-01, -2.3497e-01,
         -5.8059e-01,  1.4155e+00,  1.1241e+00, -8.7430e-01, -1.3307e+00],
        [ 1.1377e+00,  3.5718e-01,  1.4732e+00, -2.2605e-01,  5.4620e-01,
         -3.1961e-01, -2.2507e-01, -1.5210e+00, -4.4925e-01, -2.3442e-01],
        [ 3.4726e-01, -6.2492e-01,  5.8816e-01,  1.1174e+00,  1.0310e+00,
          5.3202e-01,  2.3731e-01,  5.0414e-01, -4.2485e-01, -1.8069e+00],
        [ 7.0126e-01, -1.2281e+00, -4.7729e-01,  5.2481e-01,  1.4993e+00,
          5.6660e-01,  3.9388e-01,  1.8554e-01,  5.8759e-01,  1.9689e+00],
        [-1.1635e+00,  6.4588e-01,  3.9377e-01, -5.4564e-01,  2.3864e+00,
          4.6088e-01, -8.7527e-0

In [ ]:
# Without Learnable Parameters
BatchNorm2d_layer = nn.BatchNorm2d(10, affine=False)
# With Learnable Parameters
BatchNorm2d_layer = nn.BatchNorm2d(10)
input = torch.randn(20, 10, 35, 45)
output = BatchNorm2d_layer(input)
print(input)
print(output)

tensor([[[[ 3.2498e+00, -1.1361e+00, -1.1437e+00,  ..., -1.5524e-01,
           -1.2541e+00,  2.3745e-01],
          [-1.8886e-01,  1.0084e+00,  1.1396e+00,  ...,  1.9136e+00,
            2.6322e-02,  1.0527e+00],
          [ 6.4040e-01,  7.0705e-01,  8.4769e-02,  ..., -6.8769e-01,
            2.4693e-01, -2.0326e-01],
          ...,
          [ 1.8197e+00, -4.5865e-01,  1.7775e+00,  ..., -1.9022e-01,
           -1.0356e-01, -2.0211e-01],
          [-9.5955e-01, -2.6303e-01,  4.4190e-01,  ..., -2.0048e+00,
           -9.1773e-01,  1.8528e+00],
          [ 2.0193e+00,  2.2863e-01,  3.6503e-01,  ..., -1.4027e+00,
           -7.3854e-01, -8.8414e-01]],

         [[-1.3142e+00, -4.2268e-01,  3.2736e+00,  ..., -3.4826e-01,
            1.2777e+00, -3.6433e+00],
          [-8.1349e-01, -5.1899e-01,  3.0408e-01,  ...,  1.4168e+00,
            4.9975e-02,  3.7498e-02],
          [ 4.8382e-01, -7.7343e-02, -5.5220e-02,  ...,  5.5978e-01,
            8.6317e-01, -5.6798e-01],
          ...,
     

In [ ]:
# Modified based on https://github.com/pytorch/examples/tree/master/mnist

from __future__ import print_function
import argparse
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
from torch.optim.lr_scheduler import StepLR


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, 3, 1)
        self.conv2 = nn.Conv2d(32, 64, 3, 1)
        self.BatchNorm2d=nn.BatchNorm2d(64) ###
        #self.dropout1 = nn.Dropout(0.25)
        #self.dropout2 = nn.Dropout(0.5)
        #self.BatchNorm1d_2=nn.BatchNorm1d(9216)
        self.fc1 = nn.Linear(9216, 128)
        self.BatchNorm1d=nn.BatchNorm1d(128) ###
        self.fc2 = nn.Linear(128, 10)

    def forward(self, x):
        x = self.conv1(x)
        x = F.relu(x)
        x = self.conv2(x)
        x = self.BatchNorm2d(x)
        x = F.relu(x)
        x = F.max_pool2d(x, 2)
       # x = self.dropout1(x)
        x = torch.flatten(x, 1)
        x = self.fc1(x)
        x = F.relu(x)
       # x = self.dropout2(x)
        x = self.BatchNorm1d(x)
        x = self.fc2(x)
        output = F.log_softmax(x, dim=1)
        return output


def train(args, model, device, train_loader, optimizer, epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        if batch_idx % 10 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
            epoch, batch_idx * len(data), len(train_loader.dataset),
            100. * batch_idx / len(train_loader), loss.item()))


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))



In [ ]:
    torch.manual_seed(112)
    device =torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")

In [ ]:
    transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
    dataset_train = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
    dataset_test = datasets.MNIST('../data', train=False,
                       transform=transform)

Extracting ../data/MNIST/raw/train-images-idx3-ubyte.gz to ../data/MNIST/raw



Extracting ../data/MNIST/raw/train-labels-idx1-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-images-idx3-ubyte.gz to ../data/MNIST/raw


Extracting ../data/MNIST/raw/t10k-labels-idx1-ubyte.gz to ../data/MNIST/raw
Processing...
Done!


/usr/local/lib/python3.6/dist-packages/torchvision/datasets/mnist.py:480: UserWarning: The given NumPy array is not writeable, and PyTorch does not support non-writeable tensors. This means you can write to the underlying (supposedly non-writeable) NumPy array using the tensor. You may want to copy the array to protect its data or make it writeable before converting it to a tensor. This type of warning will be suppressed for the rest of this program. (Triggered internally at  /pytorch/torch/csrc/utils/tensor_numpy.cpp:141.)
  return torch.from_numpy(parsed.astype(m[2], copy=False)).view(*s)


In [ ]:
dataset_train, dataset_test

(Dataset MNIST
     Number of datapoints: 60000
     Root location: ../data
     Split: Train
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ), Dataset MNIST
     Number of datapoints: 10000
     Root location: ../data
     Split: Test
     StandardTransform
 Transform: Compose(
                ToTensor()
                Normalize(mean=(0.1307,), std=(0.3081,))
            ))

In [ ]:
torch.manual_seed(123)
transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
        ])
dataset1 = datasets.MNIST('../data', train=True, download=True,
                       transform=transform)
dataset2 = datasets.MNIST('../data', train=False,
                       transform=transform)

train_loader = torch.utils.data.DataLoader(dataset_train, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)
test_loader = torch.utils.data.DataLoader(dataset_test, batch_size=256 , shuffle=True,
        num_workers=1,pin_memory=True)

model = Net().to(device)
optimizer = optim.Adadelta(model.parameters(), lr=0.01)

scheduler = StepLR(optimizer, step_size=5, gamma=0.1 )
for epoch in range(1, 20 + 1):
  train(10, model, device, train_loader, optimizer, epoch)
  test(model, device, test_loader)
  scheduler.step()

torch.save(model.state_dict(), "mnist_1.pt")





Train Epoch: 1 [0/60000 (0%)]	Loss: 2.395790
Train Epoch: 1 [2560/60000 (4%)]	Loss: 1.134064
Train Epoch: 1 [5120/60000 (9%)]	Loss: 0.779052
Train Epoch: 1 [7680/60000 (13%)]	Loss: 0.663981
Train Epoch: 1 [10240/60000 (17%)]	Loss: 0.574635
Train Epoch: 1 [12800/60000 (21%)]	Loss: 0.552791
Train Epoch: 1 [15360/60000 (26%)]	Loss: 0.471058
Train Epoch: 1 [17920/60000 (30%)]	Loss: 0.451472
Train Epoch: 1 [20480/60000 (34%)]	Loss: 0.393252
Train Epoch: 1 [23040/60000 (38%)]	Loss: 0.381005
Train Epoch: 1 [25600/60000 (43%)]	Loss: 0.357808
Train Epoch: 1 [28160/60000 (47%)]	Loss: 0.335644
Train Epoch: 1 [30720/60000 (51%)]	Loss: 0.302188
Train Epoch: 1 [33280/60000 (55%)]	Loss: 0.321290
Train Epoch: 1 [35840/60000 (60%)]	Loss: 0.285373
Train Epoch: 1 [38400/60000 (64%)]	Loss: 0.339933
Train Epoch: 1 [40960/60000 (68%)]	Loss: 0.240519
Train Epoch: 1 [43520/60000 (72%)]	Loss: 0.238403
Train Epoch: 1 [46080/60000 (77%)]	Loss: 0.257119
Train Epoch: 1 [48640/60000 (81%)]	Loss: 0.215105
Train E